File: app/src/main/java/com/antyzero/smoksmog/logger/CrashlyticsLogger.java
package com.antyzero.smoksmog.logger;

import android.util.Log;

import com.crashlytics.android.core.CrashlyticsCore;

import smoksmog.logger.Logger;

/**
 * Crashlytics logger
 */
public class CrashlyticsLogger implements Logger {

    private final static ConfigurationCallback EMPTY_CALLBACK = instance -> {};

    private final ExceptionLevel logExceptionLevel;

    public CrashlyticsLogger() {
        this( ExceptionLevel.VERBOSE, EMPTY_CALLBACK);
    }

    public CrashlyticsLogger(ExceptionLevel logExceptionLevel, ConfigurationCallback callback ) {
        this.logExceptionLevel = logExceptionLevel;
        callback.onConfiguration(CrashlyticsCore.getInstance());
    }

    @Override
    public void v( String tag, String message ) {
        CrashlyticsCore.getInstance().log( Log.VERBOSE, tag, message );
    }

    @Override
    public void v( String tag, String message, Throwable throwable ) {
        CrashlyticsCore.getInstance().log( Log.VERBOSE, tag, message );
        if ( logExceptionLevel.value <= ExceptionLevel.VERBOSE.value ) {
            CrashlyticsCore.getInstance().logException( throwable );
        }
    }

    @Override
    public void d( String tag, String message ) {
        CrashlyticsCore.getInstance().log( Log.DEBUG, tag, message );
    }

    @Override
    public void d( String tag, String message, Throwable throwable ) {
        CrashlyticsCore.getInstance().log( Log.DEBUG, tag, message );
        if ( logExceptionLevel.value <= ExceptionLevel.DEBUG.value ) {
            CrashlyticsCore.getInstance().logException( throwable );
        }
    }

    @Override
    public void i( String tag, String message ) {
        CrashlyticsCore.getInstance().log( Log.INFO, tag, message );
    }

    @Override
    public void i( String tag, String message, Throwable throwable ) {
        CrashlyticsCore.getInstance().log( Log.INFO, tag, message );
        if ( logExceptionLevel.value <= ExceptionLevel.INFO.value ) {
            CrashlyticsCore.getInstance().logException( throwable );
        }
    }

    @Override
    public void w( String tag, String message ) {
        CrashlyticsCore.getInstance().log( Log.WARN, tag, message );
    }

    @Override
    public void w( String tag, String message, Throwable throwable ) {
        CrashlyticsCore.getInstance().log( Log.WARN, tag, message );
        if ( logExceptionLevel.value <= ExceptionLevel.WARN.value ) {
            CrashlyticsCore.getInstance().logException( throwable );
        }
    }

    @Override
    public void e( String tag, String message ) {
        CrashlyticsCore.getInstance().log( Log.ERROR, tag, message );
    }

    @Override
    public void e( String tag, String message, Throwable throwable ) {
        CrashlyticsCore.getInstance().log( Log.ERROR, tag, message );
        if ( logExceptionLevel.value <= ExceptionLevel.ERROR.value ) {
            CrashlyticsCore.getInstance().logException( throwable );
        }
    }

    public enum ExceptionLevel {
        VERBOSE( 0 ), DEBUG( 1 ), INFO( 2 ), WARN( 3 ), ERROR( 4 );

        private final int value;

        ExceptionLevel( int value ) {
            this.value = value;
        }
    }

    public interface ConfigurationCallback {
        void onConfiguration(CrashlyticsCore instance);
    }
}

File: app/src/main/kotlin/com/antyzero/smoksmog/logger/CrashlyticsLogger.kt
package com.antyzero.smoksmog.logger

import android.util.Log

import com.crashlytics.android.core.CrashlyticsCore

import smoksmog.logger.Logger

/**
 * Crashlytics logger
 */
class CrashlyticsLogger @JvmOverloads constructor(
        private val logExceptionLevel: CrashlyticsLogger.ExceptionLevel = CrashlyticsLogger.ExceptionLevel.VERBOSE,
        callback: CrashlyticsLogger.ConfigurationCallback = CrashlyticsLogger.EMPTY_CALLBACK) : Logger {

    init {
        callback.onConfiguration(CrashlyticsCore.getInstance())
    }

    override fun v(tag: String, message: String) {
        CrashlyticsCore.getInstance().log(Log.VERBOSE, tag, message)
    }

    override fun v(tag: String, message: String, throwable: Throwable) {
        CrashlyticsCore.getInstance().log(Log.VERBOSE, tag, message)
        if (logExceptionLevel.value <= ExceptionLevel.VERBOSE.value) {
            CrashlyticsCore.getInstance().logException(throwable)
        }
    }

    override fun d(tag: String, message: String) {
        CrashlyticsCore.getInstance().log(Log.DEBUG, tag, message)
    }

    override fun d(tag: String, message: String, throwable: Throwable) {
        CrashlyticsCore.getInstance().log(Log.DEBUG, tag, message)
        if (logExceptionLevel.value <= ExceptionLevel.DEBUG.value) {
            CrashlyticsCore.getInstance().logException(throwable)
        }
    }

    override fun i(tag: String, message: String) {
        CrashlyticsCore.getInstance().log(Log.INFO, tag, message)
    }

    override fun i(tag: String, message: String, throwable: Throwable) {
        CrashlyticsCore.getInstance().log(Log.INFO, tag, message)
        if (logExceptionLevel.value <= ExceptionLevel.INFO.value) {
            CrashlyticsCore.getInstance().logException(throwable)
        }
    }

    override fun w(tag: String, message: String) {
        CrashlyticsCore.getInstance().log(Log.WARN, tag, message)
    }

    override fun w(tag: String, message: String, throwable: Throwable) {
        CrashlyticsCore.getInstance().log(Log.WARN, tag, message)
        if (logExceptionLevel.value <= ExceptionLevel.WARN.value) {
            CrashlyticsCore.getInstance().logException(throwable)
        }
    }

    override fun e(tag: String, message: String) {
        CrashlyticsCore.getInstance().log(Log.ERROR, tag, message)
    }

    override fun e(tag: String, message: String, throwable: Throwable) {
        CrashlyticsCore.getInstance().log(Log.ERROR, tag, message)
        if (logExceptionLevel.value <= ExceptionLevel.ERROR.value) {
            CrashlyticsCore.getInstance().logException(throwable)
        }
    }

    enum class ExceptionLevel constructor(internal val value: Int) {
        VERBOSE(0), DEBUG(1), INFO(2), WARN(3), ERROR(4)
    }

    interface ConfigurationCallback {
        fun onConfiguration(instance: CrashlyticsCore)
    }

    companion object {

        private val EMPTY_CALLBACK = object : ConfigurationCallback {
            override fun onConfiguration(instance: CrashlyticsCore) {

            }
        }
    }
}
